cmd/zfs/zfs_main.c

```diff
@@ -7195,14 +7233,34 @@ share_mount(int op, int argc, char **argv)
 			argv++;
 		}
 
-		if (argc != 0) {
-			(void) fprintf(stderr, gettext("too many arguments\n"));
+		if (!recursive && argc > 1) {
+			(void) fprintf(stderr, gettext("usage: "
+			    "zfs mount -a [filesystem]\n"));
 			usage(B_FALSE);
 		}
 
+		if (recursive && argc != 1) {
+			(void) fprintf(stderr, gettext("usage: "
+			    "zfs mount -r <filesystem>\n"));
+			usage(B_FALSE);
```

Comment:
This change should be dropped; it makes the usage text redundant and isn't necessary. Keep the "too many arguments" and let the main usage routine print the actual usage.